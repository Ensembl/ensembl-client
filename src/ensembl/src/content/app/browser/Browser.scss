@import 'src/styles/common';

.browser {
  width: 100%;
}

.browserInnerWrapper {
  @include flex;
}

.browserImageWrapper {
  &.semiExpanded {
    width: calc(100vw - 356px);
  }

  &.expanded {
    width: calc(100vw - 36px);
  }

  &.collapsed {
    width: 41px;

    .browserImage {
      display: none;
    }
  }
}
