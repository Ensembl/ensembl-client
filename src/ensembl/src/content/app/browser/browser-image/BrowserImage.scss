.browserImagePlus {
  height: 100%;
  position: relative;
}

.browserStage {
  height: 100%;
  width: 100%;
  overflow: hidden;
  z-index: -1;

  &.shorter {
    height: calc(100% - 56px);
  }
}

.browserInnerWrapper {
  overflow: hidden;
}

.loaderWrapper {
  left: calc(50% - 20px);
  top: calc(50% - 20px);
  position: absolute;
}
