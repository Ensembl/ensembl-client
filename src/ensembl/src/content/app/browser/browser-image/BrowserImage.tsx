import React, {
  FunctionComponent,
  RefObject,
  useEffect,
  useCallback
} from 'react';
import { connect } from 'react-redux';

import styles from './BrowserImage.scss';
import { ChrLocation, BrowserNavStates, CogList } from '../browserState';
import BrowserCogList from '../BrowserCogList';
import {
  getTrackConfigNames,
  getTrackConfigLabel,
  getBrowserCogTrackList,
  getChrLocation,
  getBrowserNavOpened,
  getBrowserActivated
} from '../browserSelectors';
import {
  activateBrowser,
  updateBrowserActivated,
  updateBrowserNavStates,
  updateChrLocation
} from '../browserActions';

import { CircleLoader } from 'src/shared/loader/Loader';

import { RootState } from 'src/store';

type StateProps = {
  browserCogTrackList: CogList;
  browserNavOpened: boolean;
  chrLocation: ChrLocation;
  trackConfigNames: any;
  trackConfigLabel: any;
  browserActivated: boolean;
};

type DispatchProps = {
  activateBrowser: (browserEl: HTMLDivElement) => void;
  updateBrowserNavStates: (browserNavStates: BrowserNavStates) => void;
  updateChrLocation: (chrLocation: ChrLocation) => void;
  updateBrowserActivated: (browserActivated: boolean) => void;
};

type OwnProps = {
  browserRef: RefObject<HTMLDivElement>;
};

type BrowserImageProps = StateProps & DispatchProps & OwnProps;

type BpaneOutEvent = Event & {
  detail: {
    bumper?: BrowserNavStates;
    location?: ChrLocation;
  };
};

export const BrowserImage: FunctionComponent<BrowserImageProps> = (
  props: BrowserImageProps
) => {
  const listenBpaneOut = useCallback((event: Event) => {
    const bpaneOutEvent = event as BpaneOutEvent;
    const navIconStates = bpaneOutEvent.detail.bumper as BrowserNavStates;
    const chrLocation = bpaneOutEvent.detail.location as ChrLocation;

    if (navIconStates) {
      props.updateBrowserNavStates(navIconStates);
    }

    if (chrLocation) {
      props.updateChrLocation(chrLocation);
    }
  }, []);

  useEffect(() => {
    const currentEl: HTMLDivElement = props.browserRef
      .current as HTMLDivElement;

    bootstrapBrowser(currentEl, props);

    currentEl.addEventListener('bpane-out', listenBpaneOut);

    return function cleanup() {
      if (currentEl && currentEl.ownerDocument) {
        props.updateBrowserActivated(false);

        currentEl.removeEventListener('bpane-out', listenBpaneOut);
      }
    };
  }, [props.browserRef]);

  useEffect(() => {
    if (props.browserCogTrackList) {
      const ons: string[] = [];
      const offs: string[] = [];

      /* what the frontend and backend call labels and names is flipped */
      Object.keys(props.browserCogTrackList).map((name) => {
        /* undefined means not seen means on for names */
        if (props.trackConfigNames[name]) {
          ons.push(name + ':label');
        } else {
          offs.push(name + ':label');
        }
        /* undefined means not seen means off for labels */
        if (props.trackConfigLabel[name] !== false) {
          ons.push(name + ':names');
        } else {
          offs.push(name + ':names');
        }
      });
      const stateEvent = new CustomEvent('bpane', {
        bubbles: true,
        detail: {
          off: offs,
          on: ons
        }
      });
      if (props.browserRef.current) {
        props.browserRef.current.dispatchEvent(stateEvent);
      }
    }
  }, [
    props.trackConfigNames,
    props.trackConfigLabel,
    props.browserRef,
    props.browserCogTrackList
  ]);

  return (
    <>
      {!props.browserActivated && (
        <div className={styles.loaderWrapper}>
          <CircleLoader />
        </div>
      )}
      <div className={styles.browserImagePlus}>
        <div
          className={getBrowserImageClasses(props.browserNavOpened)}
          ref={props.browserRef}
        />
        <BrowserCogList browserRef={props.browserRef} />
      </div>
    </>
  );
};

function bootstrapBrowser(currentEl: HTMLDivElement, props: BrowserImageProps) {
  if (currentEl && currentEl.ownerDocument) {
    const bodyEl = currentEl.ownerDocument.body as HTMLBodyElement;

    // no need to check for DOM mutations if the browser class is already set in body
    if (bodyEl.classList.contains('browser-app-ready')) {
      props.activateBrowser(currentEl);
      return;
    }

    const observerConfig = {
      attributeFilter: ['class'],
      attributes: true,
      subtree: false
    };

    const observerCallback = (mutationsList: MutationRecord[]) => {
      for (const mutation of mutationsList) {
        const mutationNode = mutation.target as HTMLElement;

        if (mutationNode.classList.contains('browser-app-ready')) {
          props.activateBrowser(currentEl);

          observer.disconnect();
          break;
        }
      }
    };

    const observer = new MutationObserver(observerCallback);

    observer.observe(bodyEl, observerConfig);
  }
}

function getBrowserImageClasses(browserNavOpened: boolean): string {
  let classes = styles.browserStage;

  if (browserNavOpened === true) {
    classes += ` ${styles.shorter}`;
  }

  return classes;
}

const mapStateToProps = (state: RootState): StateProps => ({
  browserCogTrackList: getBrowserCogTrackList(state),
  browserNavOpened: getBrowserNavOpened(state),
  chrLocation: getChrLocation(state),
  trackConfigLabel: getTrackConfigLabel(state),
  trackConfigNames: getTrackConfigNames(state),
  browserActivated: getBrowserActivated(state)
});

const mapDispatchToProps: DispatchProps = {
  activateBrowser,
  updateBrowserActivated,
  updateBrowserNavStates,
  updateChrLocation
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(BrowserImage);
