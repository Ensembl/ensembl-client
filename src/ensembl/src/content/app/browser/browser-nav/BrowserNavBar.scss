@import 'src/styles/common';

.browserNavBar {
  @include flex;
  background: $ens-light-grey;
  padding: 16px;

  @include breakpoint(large) {
    width: calc(100vw - 356px);
  }

  &.browserNavBarExpanded {
    width: 100vw;
  }
}
