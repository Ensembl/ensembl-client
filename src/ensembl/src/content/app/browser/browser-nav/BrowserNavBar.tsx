import React, { FunctionComponent, RefObject } from 'react';
import { connect } from 'react-redux';

import { browserNavConfig, BrowserNavItem } from '../browserConfig';

import { RootState } from 'src/store';
import { getBrowserNavStates } from '../browserSelectors';
import { BrowserNavStates } from '../browserState';

import BrowserNavIcon from './BrowserNavIcon';

import styles from './BrowserNavBar.scss';

type StateProps = {
  browserNavStates: BrowserNavStates;
};

type DispatchProps = {};

type OwnProps = {
  browserRef: RefObject<HTMLDivElement>;
};

type BrowserNavBarProps = StateProps & DispatchProps & OwnProps;

export const BrowserNavBar: FunctionComponent<BrowserNavBarProps> = (
  props: BrowserNavBarProps
) => {
  const browserImageEl = props.browserRef.current as HTMLDivElement;

  return (
    <div className={styles.browserNavBar}>
      <dl>
        {browserNavConfig.map((item: BrowserNavItem, index: number) => (
          <BrowserNavIcon
            key={item.name}
            browserNavItem={item}
            browserImageEl={browserImageEl}
            maxState={props.browserNavStates[index]}
          />
        ))}
      </dl>
    </div>
  );
};

const mapStateToProps = (state: RootState): StateProps => ({
  browserNavStates: getBrowserNavStates(state)
});

const mapDispatchToProps: DispatchProps = {};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(BrowserNavBar);
