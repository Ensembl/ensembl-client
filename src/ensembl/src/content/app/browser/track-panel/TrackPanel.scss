@import 'src/styles/common';

.trackPanel {
  @include flex;
  border-left: 1px solid $ens-medium-grey;
}
