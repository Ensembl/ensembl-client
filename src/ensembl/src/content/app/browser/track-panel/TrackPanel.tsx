import React, { Component } from 'react';
import { connect } from 'react-redux';

import TrackPanelBar from './track-panel-bar/TrackPanelBar';
import TrackPanelList from './track-panel-list/TrackPanelList';

import { RootState } from 'src/rootReducer';

import {
  toggleTrackPanel,
  changeCurrentTrack,
  openDrawer,
  closeDrawer
} from '../browserActions';
import {
  getCurrentTrack,
  getDrawerOpened,
  getTrackPanelOpened
} from '../browserSelectors';

import { getLaunchbarExpanded } from 'src/header/headerSelectors';

import styles from './TrackPanel.scss';

type TrackPanelProps = {
  changeCurrentTrack: (currentTrack: string) => void;
  closeDrawer: () => void;
  currentTrack: string;
  drawerOpened: boolean;
  launchbarExpanded: boolean;
  openDrawer: () => void;
  toggleTrackPanel: () => void;
  trackPanelOpened: boolean;
};

class TrackPanel extends Component<TrackPanelProps> {
  public render() {
    return (
      <section className={`${styles.trackPanel} reactSlideDrawer`}>
        <TrackPanelBar
          closeDrawer={this.props.closeDrawer}
          drawerOpened={this.props.drawerOpened}
          launchbarExpanded={this.props.launchbarExpanded}
          trackPanelOpened={this.props.trackPanelOpened}
          toggleTrackPanel={this.props.toggleTrackPanel}
        />
        {this.props.trackPanelOpened ? (
          <TrackPanelList
            currentTrack={this.props.currentTrack}
            openDrawer={this.props.openDrawer}
            updateTrack={this.props.changeCurrentTrack}
          />
        ) : null}
      </section>
    );
  }
}

const mapStateToProps = (state: RootState) => ({
  currentTrack: getCurrentTrack(state),
  drawerOpened: getDrawerOpened(state),
  launchbarExpanded: getLaunchbarExpanded(state),
  trackPanelOpened: getTrackPanelOpened(state)
});

const mapDispatchToProps = {
  changeCurrentTrack,
  closeDrawer,
  openDrawer,
  toggleTrackPanel
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TrackPanel);
