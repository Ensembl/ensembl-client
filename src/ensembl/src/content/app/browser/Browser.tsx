import React, { Component } from 'react';
import { hot } from 'react-hot-loader';
import { connect } from 'react-redux';

import BrowserBar from './browser-bar/BrowserBar';
import BrowserImage from './browser-image/BrowserImage';
import BrowserNavBar from './browser-nav/BrowserNavBar';
import TrackPanel from './track-panel/TrackPanel';
import Drawer from './drawer/Drawer';

import { RootState } from 'src/rootReducer';
import { BrowserOpenState } from './browserState';
import { closeDrawer, toggleBrowserNav } from './browserActions';
import {
  getBrowserOpenState,
  getDrawerOpened,
  getBrowserNavOpened
} from './browserSelectors';

import 'static/browser/browser';

import styles from './Browser.scss';

type BrowserProps = {
  browserNavOpened: boolean;
  browserOpenState: BrowserOpenState;
  closeDrawer: () => void;
  drawerOpened: boolean;
  toggleBrowserNav: () => void;
};

export class Browser extends Component<BrowserProps> {
  constructor(props: BrowserProps) {
    super(props);

    this.closeTrack = this.closeTrack.bind(this);
  }

  public closeTrack() {
    if (this.props.drawerOpened === false) {
      return;
    }

    this.props.closeDrawer();
  }

  public render() {
    return (
      <section className={styles.browser}>
        <BrowserBar
          browserNavOpened={this.props.browserNavOpened}
          expanded={true}
          toggleBrowserNav={this.props.toggleBrowserNav}
        />
        <div className={styles.browserInnerWrapper}>
          <div
            className={`${styles.browserImageWrapper} ${
              styles[this.props.browserOpenState]
            }`}
            onClick={this.closeTrack}
          >
            {this.props.browserNavOpened && <BrowserNavBar />}
            <BrowserImage />
          </div>
          <TrackPanel />
          {this.props.drawerOpened && <Drawer />}
        </div>
      </section>
    );
  }
}

const mapStateToProps = (state: RootState) => ({
  browserNavOpened: getBrowserNavOpened(state),
  browserOpenState: getBrowserOpenState(state),
  drawerOpened: getDrawerOpened(state)
});

const mapDispatchToProps = {
  closeDrawer,
  toggleBrowserNav
};

export default hot(module)(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(Browser)
);
