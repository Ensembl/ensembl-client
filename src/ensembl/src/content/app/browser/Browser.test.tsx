/**
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { MemoryRouter } from 'react-router';
import { mount } from 'enzyme';
import faker from 'faker';

import { BreakpointWidth } from 'src/global/globalConfig';

import { Browser, BrowserProps, ExampleObjectLinks } from './Browser';
import BrowserImage from './browser-image/BrowserImage';
import TrackPanel from './track-panel/TrackPanel';
import BrowserNavBar from './browser-nav/BrowserNavBar';

import { createChrLocationValues } from 'tests/fixtures/browser';

jest.mock('./hooks/useBrowserRouting', () => () => ({
  changeGenomeId: jest.fn()
}));
jest.mock('./browser-bar/BrowserBar', () => () => <div>BrowserBar</div>);
jest.mock('./browser-image/BrowserImage', () => () => <div>BrowserImage</div>);
jest.mock('./browser-nav/BrowserNavBar', () => () => <div>BrowserNavBar</div>);
jest.mock('./track-panel/TrackPanel', () => () => <div>TrackPanel</div>);
jest.mock('./browser-app-bar/BrowserAppBar', () => () => (
  <div>BrowserAppBar</div>
));
jest.mock('./track-panel/track-panel-bar/TrackPanelBar', () => () => (
  <div>TrackPanelBar</div>
));
jest.mock('./track-panel/track-panel-tabs/TrackPanelTabs', () => () => (
  <div>TrackPanelTabs</div>
));
jest.mock('./drawer/Drawer', () => () => <div>Drawer</div>);
jest.mock('ensembl-genome-browser', () => {
  return;
});

jest.mock('@apollo/client', () => {
  const originalModule = jest.requireActual('@apollo/client');
  return {
    ...originalModule,
    ApolloClient: jest.fn(() => {
      return {};
    })
  };
});

describe('<Browser />', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  const defaultProps: BrowserProps = {
    activeGenomeId: faker.lorem.words(),
    activeEnsObjectId: faker.lorem.words(),
    browserActivated: false,
    browserNavOpened: false,
    browserQueryParams: {},
    chrLocation: createChrLocationValues().tupleValue,
    isDrawerOpened: false,
    isTrackPanelOpened: false,
    exampleEnsObjects: [],
    toggleTrackPanel: jest.fn(),
    toggleDrawer: jest.fn(),
    viewportWidth: BreakpointWidth.DESKTOP
  };

  const wrappingComponent = (props: any) => (
    <MemoryRouter>{props.children}</MemoryRouter>
  );

  const mountBrowserComponent = (props?: Partial<BrowserProps>) =>
    mount(<Browser {...defaultProps} {...props} />, { wrappingComponent });

  describe('rendering', () => {
    test('does not render when no activeGenomeId', () => {
      const wrapper = mountBrowserComponent({ activeGenomeId: null });
      expect(wrapper.html()).toBe(null);
    });

    test('renders links to example objects only if there is no selected focus feature', () => {
      const wrapper = mountBrowserComponent();
      expect(wrapper.find(ExampleObjectLinks)).toHaveLength(1);

      wrapper.setProps({
        browserQueryParams: {
          focus: faker.lorem.words()
        }
      });
      expect(wrapper.find(ExampleObjectLinks)).toHaveLength(0);
    });

    test('renders the genome browser and track panel only when there is a selected focus feature', () => {
      const wrapper = mountBrowserComponent();

      expect(wrapper.find(BrowserImage)).toHaveLength(0);
      expect(wrapper.find(TrackPanel)).toHaveLength(0);

      wrapper.setProps({
        browserQueryParams: { focus: faker.lorem.words() }
      });

      expect(wrapper.find(BrowserImage)).toHaveLength(1);
      expect(wrapper.find(TrackPanel)).toHaveLength(1);
    });

    describe('BrowserNavBar', () => {
      const props = {
        ...defaultProps,
        browserActivated: true,
        browserQueryParams: { focus: 'foo' }
      };

      it('is rendered when props.browserNavOpened is true', () => {
        const wrapper = mountBrowserComponent(props);
        expect(wrapper.find(BrowserNavBar).length).toBe(0);

        wrapper.setProps({ browserNavOpened: true });
        expect(wrapper.find(BrowserNavBar).length).toBe(1);
      });

      it('is not rendered if drawer is opened', () => {
        const wrapper = mountBrowserComponent({
          ...props,
          browserNavOpened: true
        });
        expect(wrapper.find(BrowserNavBar).length).toBe(1);

        wrapper.setProps({ isDrawerOpened: true });
        expect(wrapper.find(BrowserNavBar).length).toBe(0);
      });
    });
  });
});
