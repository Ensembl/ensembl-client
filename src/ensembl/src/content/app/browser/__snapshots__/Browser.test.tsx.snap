// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Browser /> renders correctly 1`] = `
<section
  className="browser"
>
  <React.Memo
    browserNavOpened={true}
    expanded={true}
    toggleBrowserNav={[MockFunction]}
  />
  <div
    className="browserInnerWrapper"
  >
    <div
      className="browserImageWrapper semiExpanded"
      onClick={[Function]}
    >
      <BrowserNavBar />
      <BrowserImage />
    </div>
    <Connect(TrackPanel) />
  </div>
</section>
`;
